torchtune/modules/model_fusion/_fusion.py

```diff
@@ -463,18 +500,252 @@ def forward(
+            in the text sequence the encoder embedding outputs should be injected.
+        decoder_trainable (bool): whether to train or freeze the decoder. Default is False.
+        encoders_trainable (Union[bool, Dict[str, bool]]): whether to train or freeze the encoder. Use a single
+            boolean to set trainable for all encoders or a dictionary keyed by encoder names to specify trainable
+            for each encoder individually. Encoder names should match with ``encoders``. Default is False.
+        fusion_trainable (bool): whether to train the fusion parameters. Default is True.
+
+    Raises:
+        ValueError: if ``encoders`` and ``encoders_trainable`` keys do not match
+    """
+
+    def __init__(
+        self,
+        decoder: TransformerDecoder,
+        encoders: Dict[str, nn.Module],
+        encoder_tokens: Dict[str, int],
+        decoder_trainable: bool = False,
+        encoders_trainable: Union[bool, Dict[str, bool]] = False,
+        fusion_trainable: bool = True,
+    ):
+        super().__init__()
+        if encoders.keys() != encoder_tokens.keys() or (
+            not isinstance(encoders_trainable, bool)
+            and encoders.keys() != encoders_trainable.keys()
+        ):
+            raise ValueError(
+                f"Found mismatched keys in encoders and encoders_trainable. Got {encoders.keys()} and {encoders_trainable.keys()}."
+            )
+
+        self.decoder = decoder
+        self.encoders = nn.ModuleDict(encoders)
+        self.encoder_tokens = encoder_tokens
+        self.encoders_trainable = (
+            {k: encoders_trainable for k in self.encoders.keys()}
+            if isinstance(encoders_trainable, bool)
+            else encoders_trainable
+        )
+
+        # A little surgery in the decoder to give the
+        # fusion module access to control the embeddings
+        # The alternative is to pass a special tok_embeddings
+        # module into TransformerDecoder builder that does the
+        # merging there
+        self.tok_embeddings = decoder.tok_embeddings
+        decoder.tok_embeddings = nn.Identity()
+
+        self.register_state_dict_post_hook(self._state_dict_hook)
+        self.register_load_state_dict_pre_hook(
+            self._load_state_dict_hook, with_module=True
+        )
+
+        trainable_params = set()
+        for encoder, trainable in self.encoders_trainable.items():
+            if trainable:
+                trainable_params |= {
+                    f"encoders.{encoder}.{n}"
+                    for n, p in self.encoders[encoder].named_parameters()
+                }
+        if decoder_trainable:
+            trainable_params |= {
+                f"decoder.{n}" for n, p in self.decoder.named_parameters()
+            }
+        if fusion_trainable:
+            trainable_params |= set(get_fusion_params(self))
+        else:
+            trainable_params -= set(get_fusion_params(self))
```

Comment:
curious why would we need this?

Reply:
Not sure yet if we need this... once we have a better idea of the full architecture let's chat with @pbontrager to see which components need to be fusion modules